core/client/src/main/java/alluxio/client/file/UnderFileSystemFileInStream.java

```diff
@@ -186,20 +189,27 @@ private void checkIfClosed() {
    */
   private int directRead(byte[] b, int off, int len) throws IOException {
     int bytesLeft = len;
+    int bytesRead = 0;
+    int offset = off;
     while (bytesLeft > 0) {
-      ByteBuffer data = mReader.read(mAddress, mUfsFileId, mPos, bytesLeft);
+      // mPos shouldn't be modified, so keep track of our updated position
+      long currentPosition = mPos + bytesRead;
+
+      ByteBuffer data = mReader.read(mAddress, mUfsFileId, currentPosition, bytesLeft);
       if (data == null) { // No more data
-        if (bytesLeft == len) { // Did not read any bytes, at EOF
+        if (bytesRead == 0) { // Did not read any bytes, at EOF
```

Comment:
it seems a bit strange to only update mEOF but not mPos here. Why not updating them together outside this function?